core/src/main/java/jenkins/model/queue/CompositeCauseOfBlockage.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2016 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.model.queue;
+
+import hudson.model.TaskListener;
+import hudson.model.queue.CauseOfBlockage;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import org.apache.commons.lang.StringUtils;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Represents the fact that there was at least one {@link hudson.model.Queue.JobOffer} which rejected a task.
+ */
+@Restricted(NoExternalUse.class)
+public class CompositeCauseOfBlockage extends CauseOfBlockage {
+
+    public final Map<String, CauseOfBlockage> uniqueReasons;
+
+    public CompositeCauseOfBlockage(List<CauseOfBlockage> delegates) {
+        uniqueReasons = new TreeMap<>();
+        for (CauseOfBlockage delegate : delegates) {
+            uniqueReasons.put(delegate.getShortDescription(), delegate);
```

Comment:
It is definitely _not_ unique within a single offer, which is the whole point of keeping this map rather than a simple `List<CauseOfBlockage>`: we wish to minimize the displayed size of this reason as much as possible. For example, if 100 nodes were all considered but were all rejected because they lack a label, that becomes one map entry.